compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2213,16 +2213,20 @@ class Typer extends Namer
             traverse(xtree :: rest, facts)
           case none =>
             import untpd.modsDeco
-            val ctx1 = if (ctx.explicitNulls) {
-              mdef match {
-                case mdef: untpd.ValDef if ctx.owner.is(Method) && !mdef.mods.isOneOf(Lazy | Implicit) =>
-                  ctx.fresh.addFlowFacts(facts)
-                case _ => ctx
-              }
-            } else {
-              ctx
+            val ctx2 = mdef match {
+              case mdef: untpd.ValDef
+                if ctx.explicitNulls && ctx.owner.is(Method) && !mdef.mods.isOneOf(Lazy | Implicit) =>
```

Comment:
nit: this should move to the previous line for consistency with the rest of the code

Reply:
Regarding the condition "ctx.owner.is(Method) && !mdef.mods.isOneOf(Lazy | Implicit)", could you please double check whether forward references are allowed to vals that are lazy or implicit?

e.g.
```Scala
val y = x // is this allowed?
{implicit, lazy} val x = 42;
```

If it's not allowed, we want to get rid of the extra condition above.

Reply:
If the above are allowed, could you please add tests (there might be already) that show what the behaviour is in these cases?

Reply:
I don't think we could remove the condition `!mdef.mods.isOneOf(Lazy | Implicit)` here? If we remove it, the following test will pass:
```scala
 val x: String|Null = ???
 if (x == null) return ()
 implicit val y: Int = x.length // error: x: String|Null inferred
```

Reply:
That's fine. If the implicit is not completed earlier, then 
it's ok to infer that x is non-null.

 If the implicit is completed earlier, e.g.
```scala
val z = implicitly[int]
 val x: String|Null = ???
 if (x == null) return ()
 implicit val y: Int = x.length // error: x: String|Null inferred
```

then there should be an error.

As long as we have tests for both cases, I think it's ok to remove the check on the modifiers.}
We also need a test that shows that forward references to lazy vals triger an error.

Reply:
What if it were `lazy val z` or `def z`?

Reply:
ok, I will modify these tests next week.

Reply:
> 
> 
> What if it were `lazy val z` or `def z`?

We're already checking that it's a `ValDef`:
`case mdef: untpd.ValDef`

Yaoyu checked that forward references to lazy vals are not allowed.

Reply:
I mean if the `z` was lazy or def, not the `y`.

Oh, sorry, never mind. It can still be an error in that case.

Reply:
Yes, that would be an error because the completer is forced before we can find out that x is nullable.

The only problem I see with this scheme is that:

```scala
val x: String|Null = ???
if (x == null) return
implicit val y: Int = x.length
```

typechecks, but

```scala
val z = implicitly[int]
val x: String|Null = ???
if (x == null) return
implicit val y: Int = x.length
```

won't. The breakage will be hard to understand for the user, because it's non-local.
But that's what we get for mixing flow sensitivity with implicits.

Reply:
I'm not very confident to remove the condition of `Lazy`.  I double checked the document: https://www.scala-lang.org/files/archive/spec/2.13/04-basic-declarations-and-definitions.html.

> The scope of a name introduced by a declaration or definition is the whole statement sequence containing the binding. However, there is a restriction on forward references in blocks: In a statement sequence s1…sn making up a block, if a simple name in si refers to an entity defined by sj where j≥i, then for all sk between and including si and sj, 
> * sk cannot be a variable definition.
> * If sk is a value definition, it must be lazy.

which mean, the following program is allowed:
```scala
object Test2 {
  def ff(): Unit = {
    lazy val z = y
    lazy val x: String = ???
    if (x == "") return ()
    lazy val y = x.length
  }
}
```

Reply:
Good find! Let's leave the check for 'lazy' then, and add a comment explaining why we special case lazy vals.